openmc/lib/tally.py

```diff
@@ -405,4 +402,18 @@ def __len__(self):
     def __repr__(self):
         return repr(dict(self))
 
+    def __delitem__(self,key):
+        """Delete a tally from tally vector and remove the ID,index pair from tally"""
+        _dll.openmc_remove_tally(self._get_tally_index(key))
+
+    def _get_tally_index(self,key):
+        """Given the ID, return the index"""
+        index = c_int32()
+        try:
+            _dll.openmc_get_tally_index(key, index)
+        except (AllocationError, InvalidIDError) as e:
+            # __contains__ expects a KeyError to work correctly
+            raise KeyError(str(e))
+        return index.value
```

Comment:
I think this should work:
```suggestion
        _dll.openmc_remove_tally(self[key]._index)
```

Reply:
Ah, good call. Nice and elegant. I suggested the refactor here, but this is better.

Reply:
My Class knowledge in python is still in development. I can guess that `self[key]._index` retrieves the proper index for this function, but how does this work exactly? Also, I think because I pushed another change in this file, this is now outdated, and I can't "Commit suggestoin," for the co-author, but I will credit you in the commit.

Reply:
Just want to confirm, but if I delete `_get_tally_index` then I also need to modify `__getitem__` to be
```
    def __getitem__(self, key):
        return Tally(index=self[key])
```

Reply:
Only thing I'm concerned about is that we've now lost the `KeyError` checking

Reply:
`self[key]` is the equivalent of `self.__getitem__(key)`. The `__getitem__` method already has the logic for checking the key and will raise a `KeyError` is you pass an invalid key

Reply:
Good to know then! I will add this in and then I think it will be ready for merge, assuming everything passes and there's no new corrections.